### PR TITLE
[Stacked PR 6/6] Optimize GDN forward/backward Pallas kernels and add Head-Sharded & Sequence-Sharded Context Parallelism (CP)

### PR DESCRIPTION
**Pull request by Rohan-Bierneni**
# Description

This is **PR 6 of 6** in the stacked series establishing the canonical Pallas Gated Delta Net (GDN) implementation in MaxText.

This PR completes the GDN stack on top of PR 5 (`#5098`), consolidating the optimized forward and backward Pallas kernels and introducing full **Head-Sharded and Sequence-Sharded Context Parallelism (CP)**:

1. **True Sequence-Sharded Context Parallelism (`cp_gdn.py`)**:
   - `halo_exchange_for_conv` and `halo_exchange_for_conv_bwd`: causal Conv1D boundary halo exchange of $K - 1 = 3$ pre-conv QKV tokens via `lax.ppermute` across `cp_axis`, with cross-rank packed document `segment_ids` masking.
   - `compose_local_from_t_inv`: pre-batched forward affine transition summary using cached $T_c^{-1}$ (`t_inv`) from the forward Pallas kernel, convolving only $K$ channels (`1/6`th of QKV) and folding across local chunks via an $\mathcal{O}(\log_2 N_c)$ pairwise tree reduction.
   - 0-GEMM backward transition matrix transpose ($\mathrm{d}M_{\mathrm{loc}} = M_{\mathrm{loc}}^\top$) from cached `"gdn_m_local"` and pre-batched state adjoint summary (`compose_bwd_local_from_t_inv`) with $\mathcal{O}(\log_2 N_c)$ tree reduction.
   - `incoming_grad_state`: $\lceil \log_2(D) \rceil$-step reverse Hillis-Steele prefix scan over `cp_axis` in $\mathcal{O}(1)$ HBM per device.

2. **Unified CP Routing in `Qwen3NextGatedDeltaNet` (`src/maxtext/models/qwen3.py`)**:
   - Selectable via `gdn_cp_mode: "auto" | "seq" | "head"` (defaulting to `"auto"`).
   - `"auto"` uses Head-Sharded CP at `CP = 2` (where `num_groups: 2 -> 1` yields `103 ms` single-pass step latency) and automatically switches to Sequence-Sharded CP at `CP >= 4` (scaling to `63 ms` at `CP = 4`, `32 ms` at `CP = 8`, and `8.9 ms` at `CP = 32`).
   - Supports `cp_size > num_k_heads` (e.g. `CP = 8` with `num_k_heads = 4` and `CP = 32` with `num_k_heads = 16`).

3. **Pallas Forward & Backward Kernel Optimizations**:
   - Backward pass optimizations: native BF16 MXU execution, algebraic GEMM reductions, and multi-group grid (`(batch_size, num_groups, num_chunks)`).
   - Forward pass optimizations: fast transpose broadcasts, TPU dot safeguards, and stable triangular inversion.
   - Conditional VMEM BlockSpecs: `has_dht` and `has_dh0` are only allocated in `pallas_gdn_bwd_kernel` when initial/terminal recurrent states are present, preventing any VMEM regression in pure FP32 mode.

# Tests
- **Multi-Device CPU Suite**: All tests pass in `venv_maxtext/bin/pytest`:
  - `gdn_cp_test.py`: 7 passed (including 8-device mock scan).
  - `gdn_cp_head_sharded_test.py`: 5 passed (with 4-device emulation).
  - `gdn_bwd_pallas_test.py`: 19 passed (including `test_gdn_decoupled_conv1d_cp_axis_against_single_device` and `test_bwd_block_specs_and_symbolic_zero_dht_dh0_omission`).
  - `gdn_benchmark_test.py`: 6 passed.
  - `pyconfig_test.py`: 46 passed.
- **Hardware Verification on Cloud TPU v7x (`TPU7x`)**:
  - `4`-chip (`8` `TPU7x` TensorCores) multi-chip CP sweep (`CP = 1, 2, 4, 8` side-by-side) verified on Ghostfish hardware.
  - Full non-CP `gdn_benchmark_ghostfish_test` suite (all 6 tests, including 64K and 8K FP32/BF16) verified on Ghostfish hardware.

# Checklist
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests and verified parity across CPU and TPU.

### CODE REVIEW

File: src/maxtext/models/qwen3.py [hunk 1]
```diff
@@ -986,41 +1006,131 @@ def extract_state(c_in, v_len):
...
+            conv_bias_pspec = P()
+
+          conv_state_arg = (
+              conv_state
+              if conv_state is not None
+              else jnp.zeros(
+                  (batch, self.config.gdn_conv_kernel_dim - 1, num_groups, channels_per_group),
+                  dtype=cfg.dtype,
+              )
+          )
```
[the remark this answers is not in the document]
Comment (Rohan-Bierneni) [reply src/maxtext/models/qwen3.py]: Done. Implemented component-wise slicing and 4D grouping for incoming `conv_state`:
```python
if conv_state is not None:
  cs_q = conv_state[..., : self.key_dim].reshape(
      batch, self.config.gdn_conv_kernel_dim - 1, num_groups, self.head_k_dim
  )
  cs_k = conv_state[..., self.key_dim : 2 * self.key_dim].reshape(
      batch, self.config.gdn_conv_kernel_dim - 1, num_groups, self.head_k_dim
  )
  cs_v = conv_state[..., 2 * self.key_dim :].reshape(
      batch, self.config.gdn_conv_kernel_dim - 1, num_groups, self.v_heads_per_k_head * self.head_v_dim
  )
  conv_state_arg = jnp.concatenate([cs_q, cs_k, cs_v], axis=-1)
else:
  conv_state_arg = jnp.zeros(
      (batch, self.config.gdn_conv_kernel_dim - 1, num_groups, channels_per_group),
      dtype=cfg.dtype,
  )
